<a href="https://colab.research.google.com/github/ivanavion-collab/Phys-chem/blob/main/workshop3_thermodynamicsA_2026.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Workshop 3: Thermodynamics of Gas-Phase Acidity

##Overview
Quantum chemical calculations will be used to investigate the thermodynamics of gas-phase proton abstraction reactions to determine the relative gas-phase acidities of water and other chemical species.

## Background
Reactions of acids or bases play important roles in many areas, from biology to chemistry to geology. According to the Brønsted–Lowry definition, an acid is a proton donor and a base is a proton acceptor.<a name="cite_ref-1"></a>[<sup>[1]</sup>](#cite_note-1) A key property of an acid is its strength, or propensity to donate a proton, which is characterised by the equilibrium constant 𝐾a of the acid ionisation reaction.

HA ⇌ H$^{+}$ + A$^{-}$ &emsp; &emsp; $K_a = \frac{[\mathrm{H}^+][\mathrm{A}^-]}{[\mathrm{H}\mathrm{A}]}$ (1)

where HA is the acid and A$^-$ is its conjugate base. The higher the value of 𝐾a, the greater the propensity of the acid to give up a proton and thus the stronger the acid. The acid-ionisation equilibrium constant can be calculated from the standard Gibbs free energy change ∆𝐺° for the acid ionisation reaction using:

$ΔG^\circ = -RT\ln K_a$ (2)

The standard Gibbs free energy change ∆𝐺° of a chemical reaction is the Gibbs free energy change ∆𝐺 in the case in which each reactant and product is at 1 atm pressure (for a gas) or 1 M concentration (for a solution).

Measuring the acid-ionisation equilibrium constant can be experimentally challenging, particularly for weak acids with small equilibrium constants. Fortunately, the free energy change, and thus the acid-ionisation equilibrium constant, can be accurately calculated in many cases using quantum chemistry. This practical will use a particular type of quantum chemistry called density functional theory, DFT (as implemented in Psi4<a name="cite_ref-2"></a>[<sup>[2]</sup>](#cite_note-2)) and the atomic simulation environment<a name="cite_ref-3"></a>[<sup>[3]</sup>](#cite_note-3) to calculate the thermodynamics of acid ionisation. DFT is a method that calculates the ground-state electron density and electronic energy for a specified nuclear geometry.<a name="cite_ref-4"></a>[<sup>[4]</sup>](#cite_note-4) The calculations are for molecules in the gas phase (i.e. not solvated), which can serve as a good starting point for more sophisticated calculations of the thermodynamics of acid ionisation in solution. Furthermore, gas-phase acidity and basicity play a significant role in important chemical processes, e.g. determining the propensity to form molecular ions of particular mass/charge ratios in mass spectrometry.<a name="cite_ref-5"></a>[<sup>[5]</sup>](#cite_note-5)

By varying the nuclear geometry to minimise the electronic energy in a process called geometry optimisation, the equilibrium geometry of a molecule at absolute zero temperature (0 K) can be determined. To calculate the total energy, enthalpy, or Gibbs free energy at finite (non-zero) temperature, contributions from translational, rotational, and vibrational degrees of freedom and the effect of finite temperature must be included. By calculating the electronic energy for small displacements of the atoms from the equilibrium geometry, the oscillation frequencies of the vibrational degrees of freedom (normal modes) of the molecule can be estimated using the harmonic oscillator approximation in a process called vibrational frequency analysis, from which quantum vibrational energies can be calculated. Using the electronic energy, equilibrium geometry, and vibrational frequencies from these quantum-chemical calculations, along with the molecular weight, in equations derived from statistical mechanics, the thermodynamics at finite temperature can be calculated.

## Goals and Objectives
_Overall Goals_:
1. Learn how quantum chemical calculations can be used to investigate the geometry, electronic structure, and
thermochemistry of a simple molecular system.
2. Understand the output of quantum chemical calculations.
3. Use the theory of statistical mechanics to understand how the translational, vibrational, and electronic contributions to the average energy of an isolated molecule at a given temperature is calculated from molecular parameters and the single-molecule canonical partition function.
4. Evaluate the importance of translational, rotational, and/or vibrational contributions and/or finite-temperature corrections to the thermodynamics of chemical reactions.
5. Understand the relationship between energy, enthalpy, and Gibbs free energy.
6. Discuss the gas-phase acidity of a molecular species in terms of its molecular structure.

##Specific Objectives:
1. Build and optimise molecules using quantum chemical calculations.
2. Extract the energies, geometries and vibrational frequencies of conjugate acid–base pairs from quantum chemical calculations.
2. Determine the standard Gibbs free energy change and equilibrium constant for acid ionisation in the gas phase for several acids.
3. Compare the calculated gas-phase acidities of different acids and compare the calculated values with experimentally measured values.

---
<a name="cite_note-1"></a>1. [^](#cite_ref-1)  A. Blackman, S.E. Bottle, S. Schmid, M. Mocerino, U. Wille, Chemistry (3e), Wiley, 2015.

<a name="cite_note-2"></a>2. [^](#cite_ref-2)  https://psicode.org

<a name="cite_note-3"></a>3. [^](#cite_ref-3)  https://wiki.fysik.dtu.dk/ase/

<a name="cite_note-4"></a>4. [^](#cite_ref-5)  D.S. Sholl, J.A. Steckel, Density Functional Theory: A Practical Introducton, Wiley-Interscience, 2009.

<a name="cite_note-5"></a>5. [^](#cite_ref-5)  A.G. Harrison, Mass Spectrometry Reviews 16, 201–217 (1997)

# A. Autoionisation of Water
Combined geometry optimisation and vibrational frequency calculations will be performed for water and its conjugate base, the hydroxide ion, using DFT (at the _B3LYP/6-31+G(d)_ level of theory). Use the results of these calculations to calculate the change in energy, standard enthalpy, and standard Gibbs free energy for the gas- phase acid ionisation of water at 298.15 K (25$^\circ$C).

First we need to install the DFT codes (this will take a few minutes)

In [1]:
!pip install -q condacolab
import condacolab
condacolab.install()

⏬ Downloading https://github.com/jaimergp/miniforge/releases/download/24.11.2-1_colab/Miniforge3-colab-24.11.2-1_colab-Linux-x86_64.sh...
📦 Installing...
📌 Adjusting configuration...
🩹 Patching environment...
⏲ Done in 0:00:10
🔁 Restarting kernel...


In [1]:
!rm -f /usr/local/conda-meta/pinned
!mamba install -y python=3.11 psi4 ase -c conda-forge


Looking for: ['python=3.11', 'psi4', 'ase']

[+] 0.0s
conda-forge/linux-64  ⣾  
conda-forge/noarch    ⣾  [+] 0.1s
conda-forge/linux-64  ⣾  
conda-forge/noarch     1%[+] 0.2s
conda-forge/linux-64   2%
conda-forge/noarch     7%[+] 0.3s
conda-forge/linux-64   6%
conda-forge/noarch    15%[+] 0.4s
conda-forge/linux-64   8%
conda-forge/noarch    20%[+] 0.5s
conda-forge/linux-64  11%
conda-forge/noarch    23%[+] 0.6s
conda-forge/linux-64  15%
conda-forge/noarch    33%[+] 0.7s
conda-forge/linux-64  18%
conda-forge/noarch    40%[+] 0.8s
conda-forge/linux-64  19%
conda-forge/noarch    43%[+] 0.9s
conda-forge/linux-64  21%
conda-forge/noarch    46%[+] 1.0s
conda-forge/linux-64  23%
conda-forge/noarch    49%[+] 1.1s
conda-forge/linux-64  25%
conda-forge/noarch    51%[+] 1.2s
conda-forge/linux-64  26%
conda-forge/noarch    55%[+] 1.3s
conda-forge/linux-64  28%
conda-forge/noarch    62%[+] 1.4s
conda-forge/linux-64  31%
conda-forge/noarch    65%[+] 1.5s
conda-forge/linux-64  33%
conda-forge/noarch 

In [2]:
%%bash
# verify installation
python -c "
import psi4
import ase
print(f'Psi4 version: {psi4.__version__}')
print(f'ASE version: {ase.__version__}')
"

Psi4 version: 1.10
ASE version: 3.28.0


**Important:** Because the conda installation changes the Python version, psi4 cannot be imported directly in the Colab notebook kernel. Instead, all code that uses psi4 must be run via `%%bash` cells that execute external Python scripts.

The _ase_ (atomic simulation environment) library has routines for describing molecules and interfaces with DFT software. For example, we can either describe a molecule using a string ('H2O') or by explicitly listing the position of atoms.

In [3]:
%%bash
cat << 'EOF' > setup_molecules.py
from ase.build import molecule
from ase import Atoms
from ase.io import write
import numpy as np

water = molecule('H2O')

d=0.97
ohminus = Atoms('OH', positions=([0-d/2,0,0],[0+d/2,0,0]), charges=[-1,0])

hplus = molecule('H')
hplus.set_initial_charges([+1])

# Write xyz files for visualisation
write(filename='water.xyz', images=water)
write(filename='ohminus.xyz', images=ohminus)

print("Molecules created successfully")
print(f"Water: {water.get_chemical_formula()}")
print(f"OH-: {ohminus.get_chemical_formula()}")
print(f"H+: {hplus.get_chemical_formula()}")
EOF

python setup_molecules.py

Molecules created successfully
Water: H2O
OH-: HO
H+: H


Computing the energy for H$^{+}$ is doesn't require any geometry optimisation (it's only one atom).

The calculation parameters are loaded with the Psi4 object. Check out https://wiki.fysik.dtu.dk/ase/ase/calculators/psi4.html for the full list of options. **There is an error in the parameters below!**

In [9]:
%%bash
cat << 'EOF' > calc_hplus.py
from ase.build import molecule
from ase import Atoms
from ase.calculators.psi4 import Psi4
import json

hplus = molecule('H')
hplus.set_initial_charges([+1])

calc = Psi4(atoms = hplus,
            method = 'b3lyp',
            memory = '500MB',
            basis = '6-31+G(d)',
            charge = 1,
            multiplicity=1,
            label = hplus.get_chemical_formula())

hplus.calc = calc
energy = hplus.get_potential_energy()
print(f"H+ energy: {energy}")

# Save result to file for later use
results = {'energy': energy}
with open('h_results.json', 'w') as f:
    json.dump(results, f)
EOF

python calc_hplus.py


  Memory set to 476.837 MiB by Python driver.
  Threads set to 1 by Python driver.

  Memory set to 476.837 MiB by Python driver.
  Threads set to 1 by Python driver.
H+ energy: 0.0


/usr/local/lib/python3.11/site-packages/ase/calculators/psi4.py:127: UserWarning: A charge was provided without a spin multiplicity. A multiplicity of 1 is assumed
  warnings.warn('A charge was provided without a spin '


__What is the energy for H$^+$ and in which units?__

 Answer: 0.0 eV (2sf)

For molecules we need to first optimise the atom positions to find the lowest energy positons. The calculation optimises the molecular structure through a series of geometry optimisation steps.

In [11]:
%%bash
cat << 'EOF' > calc_geomopt.py
from ase.calculators.psi4 import Psi4
from ase.build import molecule
from ase import Atoms
import numpy as np
from ase.optimize import QuasiNewton
import json
import pickle

# Recreate molecules
water = molecule('H2O')
d=0.97
ohminus = Atoms('OH', positions=([0-d/2,0,0],[0+d/2,0,0]), charges=[-1,0])
molecules = [water, ohminus]

optimised_molecules = []
results = {}
for atoms in molecules:
  formula = atoms.get_chemical_formula()
  print(formula)
  calc = Psi4(atoms = atoms,
        method = 'b3lyp',
        memory = '500MB',
        basis = '6-31+G(d)',
        charge = int(np.sum(atoms.get_initial_charges())),
        multiplicity=1,
        label = formula)

  atoms.calc = calc

  QuasiNewton(atoms).run(fmax=0.0005)

  energy = atoms.get_potential_energy()
  if formula == 'HO':
      results['oh_energy'] = energy
  if formula == 'H2O':
      results['h2o_energy'] = energy
  print(f"{formula} energy: {energy}")

  # Remove calculator before saving (psi4 module cannot be pickled)
  atoms.calc = None
  optimised_molecules.append(atoms)

# Save results and optimised molecules for later
with open('geomopt_results.json', 'w') as f:
    json.dump(results, f)
with open('optimised_molecules.pkl', 'wb') as f:
    pickle.dump(optimised_molecules, f)

print("\nOptimisation complete. Results saved.")
EOF

python calc_geomopt.py


  Memory set to 476.837 MiB by Python driver.
  Threads set to 1 by Python driver.

  Memory set to 476.837 MiB by Python driver.
  Threads set to 1 by Python driver.
H2O
                Step[ FC]     Time          Energy          fmax
BFGSLineSearch:    0[  0] 03:52:49    -2079.562595       0.1365
BFGSLineSearch:    1[  2] 03:52:54    -2079.564058       0.0820
BFGSLineSearch:    2[  4] 03:52:59    -2079.564177       0.0003
H2O energy: -2079.564177173864
HO
                Step[ FC]     Time          Energy          fmax
BFGSLineSearch:    0[  0] 03:53:00    -2062.532144       0.2308
BFGSLineSearch:    1[  2] 03:53:05    -2062.532703       0.0012
BFGSLineSearch:    2[  3] 03:53:07    -2062.532703       0.0000
HO energy: -2062.5327029106875

Optimisation complete. Results saved.


 __What is the convergence criteria used here (Don't forget the units)?__

 Answer: 0.0005 N

  __What optimiser is used here and what alternatives are there?__

 Answer: QuasiNewtonis used in the code above.
 BFGSline force also used?
 Alternatives: FIRE, GPMin, LBFGS


The output displayed is only a snapshot of important quantities. The files _H.dat_, _HO.dat_ and _H20.dat_ have the raw output from the _Psi4_ sofware. Check the files tab on the left side of the browser.

__Open the *H2O.dat* file and search the detailed ouput. Which point group symmetry is used here?__
C1 molecular point group, full point group: Cs, H20  has c2v symetry.

Using the harmonic oscilator approximation the vibrational degrees of freedom (normal modes) of these molecules can be computed.

In [16]:
%%bash
cat << 'EOF' > calc_vibrations.py
from ase.vibrations import Vibrations
from ase.calculators.psi4 import Psi4
import numpy as np
import pickle
import json

with open('optimised_molecules.pkl', 'rb') as f:
    optimised_molecules = pickle.load(f)

vibs = []
vib_data = []
for atoms in optimised_molecules:
  formula = atoms.get_chemical_formula()
  print(formula)

  # Reattach calculator (was stripped for pickling)
  calc = Psi4(atoms = atoms,
        method = 'b3lyp',
        memory = '500MB',
        basis = '6-31+G(d)',
        charge = int(np.sum(atoms.get_initial_charges())),
        multiplicity=1,
        label = formula)
  atoms.calc = calc

  vib = Vibrations(atoms, name=formula, delta=0.01)
  vib.clean()
  vib.run()
  vib.write_mode(0)
  vib.write_mode(-1)
  vib.summary()

 # Save vib energies - convert to real+imag pairs for JSON serialisation
  # (some modes have imaginary frequencies represented as complex numbers)
  vib_energies = vib.get_energies()
  vib_energies_serialised = [{'real': float(e.real), 'imag': float(e.imag)} for e in vib_energies]
  vib_data.append({'formula': formula, 'vib_energies': vib_energies_serialised})
  vibs.append(vib)

# Save vib energies for later use
with open('vib_data.json', 'w') as f:
    json.dump(vib_data, f)

print("\nVibration analysis complete.")
EOF

MPLBACKEND=agg python calc_vibrations.py


  Memory set to 476.837 MiB by Python driver.
  Threads set to 1 by Python driver.

  Memory set to 476.837 MiB by Python driver.
  Threads set to 1 by Python driver.
H2O
---------------------
  #    meV     cm^-1
---------------------
  0    6.3i     51.1i
  1    0.3i      2.8i
  2    0.2i      1.6i
  3    0.1       0.8
  4    3.3      26.7
  5    3.4      27.6
  6  206.0    1661.8
  7  463.2    3736.1
  8  478.6    3860.4
---------------------
Zero-point energy: 0.577 eV
HO
---------------------
  #    meV     cm^-1
---------------------
  0    0.0i      0.0i
  1    0.0i      0.0i
  2    0.0i      0.0i
  3    3.3      26.7
  4    3.3      26.7
  5  453.3    3656.3
---------------------
Zero-point energy: 0.230 eV

Vibration analysis complete.


The modes as displayed include translational and rotational movements that leads low energy negative frequences, these can be ignored.

The trajectory of the 0th mode was saved (_H2O.0.traj_) download this to your local computer. You can view this using the _view_ routine in _ase__ as demonstrated in **['workshop3-view.ipynb'](https://colab.research.google.com/github/jackevansadl/CHEM3630/blob/main/workshop3-view.ipynb)**.

__What do the 8th vibrational mode of H2O and the 5th mode of HO relate to?__.
_You may need to reload the output of the **['workshop3-view.ipynb'](https://colab.research.google.com/github/jackevansadl/CHEM3630/blob/main/workshop3-view.ipynb)** notebook_

_Answer:_ these vibrational modes are asymetric stretches as the dipole moment of these molecules change.


The _thermochemistry_ routines of ASE can be used to compute several thermodynamic quantites from the normal modes using the ideal gas approximation. An example for H2O is shown below:

In [17]:
%%bash
cat << 'EOF' > calc_thermo_h2o.py
from ase.thermochemistry import IdealGasThermo
import pickle
import json
import numpy as np

with open('optimised_molecules.pkl', 'rb') as f:
    optimised_molecules = pickle.load(f)
with open('vib_data.json', 'r') as f:
    vib_data = json.load(f)
with open('geomopt_results.json', 'r') as f:
    geomopt_results = json.load(f)

# the 0th entry is h2o
atoms = optimised_molecules[0]
potentialenergy = geomopt_results['h2o_energy']

# Reconstruct complex vib energies as a plain list (IdealGasThermo expects a list)
raw = vib_data[0]['vib_energies']
vib_energies = [complex(e['real'], e['imag']) for e in raw]
print(f"vib_energies type: {type(vib_energies)}, first element: {vib_energies[0]} (type: {type(vib_energies[0])})")

thermo = IdealGasThermo(vib_energies=vib_energies,
                        potentialenergy=potentialenergy,
                        atoms=atoms,
                        geometry='linear',
                        symmetrynumber=2, spin=0)
G = thermo.get_gibbs_energy(temperature=298.15, pressure=101325.)
H = thermo.get_enthalpy(temperature=298.15, verbose=False)
S = thermo.get_entropy(temperature=298.15, pressure=101325., verbose=False)

print(f"H2O Gibbs energy: {G}")
print(f"H2O Enthalpy: {H}")
print(f"H2O Entropy: {S}")

# Save results
results = {'gibbs_energy': G, 'enthalpy': H, 'entropy': S}
with open('h2o_thermo.json', 'w') as f:
    json.dump(results, f)
EOF

MPLBACKEND=agg python calc_thermo_h2o.py

vib_energies type: <class 'list'>, first element: 0.006336255766631611j (type: <class 'complex'>)
Enthalpy components at T = 298.15 K:
Enthalpy components at T = 298.15 K:
E_pot              -2079.564 eV
E_ZPE                  0.576 eV
Cv_trans (0->T)        0.039 eV
Cv_rot (0->T)          0.026 eV
Cv_vib (0->T)          0.024 eV
-------------------------------
U                  -2078.900 eV
(C_v -> C_p)           0.026 eV
-------------------------------
H                  -2078.875 eV

Entropy components at T = 298.15 K and P = 101325.0 Pa:
{"":15s}{"S":13s}     {"T*S:13s}
S_trans (1 bar)    0.0015019 eV/K        0.448 eV
S_rot              0.0002940 eV/K        0.088 eV
S_elec             0.0000000 eV/K        0.000 eV
S_vib              0.0002601 eV/K        0.078 eV
S (1 bar -> P)    -0.0000011 eV/K       -0.000 eV
-------------------------------------------------
S                  0.0020549 eV/K        0.613 eV

Free energy components at T = 298.15 K and P = 101325.0 Pa:
    H  

__Confirm that the average translation energy (described here as translational heat capacity Cv) is described by Eq. 9 (Appendix):__

$<ɛ_T> = -\frac{1}{q_T}(\frac{∂q_T}{∂β})_{N,V} \approx \frac{3}{2}k_BT$

In [26]:


kb = 8.6173303 * 10**-5
T = 298.15
e = 3/2 * kb * T
print (e)

#answer goes here (use %%bash and a Python script if psi4/ase routines are needed)

0.038538855434175


Cv_trans (0->T)        0.039 eV
average translational energy was 0.0385 eV
therefore they are consistent with one another

__Compute the same thermodynamic quanties (gibbs energy, enthalpy and entropy) for OH$^-$ and H$^+$.__

In [29]:
%%bash
#copy cell above
#update fo OH-
#oth to 1st

cat << 'EOF' > calc_thermo_oh.py
from ase.thermochemistry import IdealGasThermo
import pickle
import json
import numpy as np


with open('optimised_molecules.pkl', 'rb') as f:
    optimised_molecules = pickle.load(f)
with open('vib_data.json', 'r') as f:
    vib_data = json.load(f)
with open('geomopt_results.json', 'r') as f:
    geomopt_results = json.load(f)

# the 1st entry is OH
atoms = optimised_molecules[1]
potentialenergy = geomopt_results['oh_energy']

# Reconstruct complex vib energies as a plain list (IdealGasThermo expects a list)
raw = vib_data[1]['vib_energies']
vib_energies = [complex(e['real'], e['imag']) for e in raw]
print(f"vib_energies type: {type(vib_energies)}, first element: {vib_energies[0]} (type: {type(vib_energies[0])})")

thermo = IdealGasThermo(vib_energies=vib_energies,
                        potentialenergy=potentialenergy,
                        atoms=atoms,
                        geometry='linear',
                        symmetrynumber=2, spin=0)
G = thermo.get_gibbs_energy(temperature=298.15, pressure=101325.)
H = thermo.get_enthalpy(temperature=298.15, verbose=False)
S = thermo.get_entropy(temperature=298.15, pressure=101325., verbose=False)

print(f"OH Gibbs energy: {G}")
print(f"OH Enthalpy: {H}")
print(f"OH Entropy: {S}")

# Save results
results = {'gibbs_energy': G, 'enthalpy': H, 'entropy': S}
with open('oh_thermo.json', 'w') as f:
    json.dump(results, f)
EOF

MPLBACKEND=agg python calc_thermo_oh.py
# answer for OH-
# Write your script here following the same pattern as the H2O thermochemistry cell above
# Load optimised_molecules.pkl, vib_data.json (index 1 for OH-), and geomopt_results.json
# Save results to oh_thermo.json

vib_energies type: <class 'list'>, first element: 3.2972521980922196e-06j (type: <class 'complex'>)
Enthalpy components at T = 298.15 K:
Enthalpy components at T = 298.15 K:
E_pot              -2062.533 eV
E_ZPE                  0.227 eV
Cv_trans (0->T)        0.039 eV
Cv_rot (0->T)          0.026 eV
Cv_vib (0->T)          0.000 eV
-------------------------------
U                  -2062.242 eV
(C_v -> C_p)           0.026 eV
-------------------------------
H                  -2062.216 eV

Entropy components at T = 298.15 K and P = 101325.0 Pa:
{"":15s}{"S":13s}     {"T*S:13s}
S_trans (1 bar)    0.0014945 eV/K        0.446 eV
S_rot              0.0002337 eV/K        0.070 eV
S_elec             0.0000000 eV/K        0.000 eV
S_vib              0.0000000 eV/K        0.000 eV
S (1 bar -> P)    -0.0000011 eV/K       -0.000 eV
-------------------------------------------------
S                  0.0017270 eV/K        0.515 eV

Free energy components at T = 298.15 K and P = 101325.0 Pa:
    H

In [38]:
%%bash

cat << 'EOF' > calc_thermo_h.py
from ase.thermochemistry import IdealGasThermo
import pickle
import json
import numpy as np
from ase.build import molecule


# the 1st entry is OH
atoms = molecule('H')
potentialenergy = 0.0



thermo = IdealGasThermo(vib_energies=[],
                        potentialenergy=potentialenergy,
                        atoms=atoms,
                        geometry='monatomic',
                        symmetrynumber=2, spin=0)
G = thermo.get_gibbs_energy(temperature=298.15, pressure=101325.)
H = thermo.get_enthalpy(temperature=298.15, verbose=False)
S = thermo.get_entropy(temperature=298.15, pressure=101325., verbose=False)

print(f"H Gibbs energy: {G}")
print(f"H Enthalpy: {H}")
print(f"H Entropy: {S}")

# Save results
results = {'gibbs_energy': G, 'enthalpy': H, 'entropy': S}
with open('h_thermo.json', 'w') as f:
    json.dump(results, f)
EOF

MPLBACKEND=agg python calc_thermo_h.py
# answer for H+ (Hint: consider the 'geometry' parameter and set vib_energies to an empty list [])
# Load h_results.json for the H+ energy
# Save results to h_thermo.json
no longer need to OPEN all that crap

Enthalpy components at T = 298.15 K:
Enthalpy components at T = 298.15 K:
E_pot                  0.000 eV
E_ZPE                  0.000 eV
Cv_trans (0->T)        0.039 eV
Cv_rot (0->T)          0.000 eV
Cv_vib (0->T)          0.000 eV
-------------------------------
U                      0.039 eV
(C_v -> C_p)           0.026 eV
-------------------------------
H                      0.064 eV

Entropy components at T = 298.15 K and P = 101325.0 Pa:
{"":15s}{"S":13s}     {"T*S:13s}
S_trans (1 bar)    0.0011292 eV/K        0.337 eV
S_rot              0.0000000 eV/K        0.000 eV
S_elec             0.0000000 eV/K        0.000 eV
S_vib              0.0000000 eV/K        0.000 eV
S (1 bar -> P)    -0.0000011 eV/K       -0.000 eV
-------------------------------------------------
S                  0.0011281 eV/K        0.336 eV

Free energy components at T = 298.15 K and P = 101325.0 Pa:
    H          0.064 eV
 -T*S         -0.336 eV
-----------------------
    G         -0.272 eV
H Gibbs e

bash: line 39: no: command not found


CalledProcessError: Command 'b'\ncat << \'EOF\' > calc_thermo_h.py\nfrom ase.thermochemistry import IdealGasThermo\nimport pickle\nimport json\nimport numpy as np\nfrom ase.build import molecule\n\n\n# the 1st entry is OH\natoms = molecule(\'H\')\npotentialenergy = 0.0\n\n\n\nthermo = IdealGasThermo(vib_energies=[],\n                        potentialenergy=potentialenergy,\n                        atoms=atoms,\n                        geometry=\'monatomic\',\n                        symmetrynumber=2, spin=0)\nG = thermo.get_gibbs_energy(temperature=298.15, pressure=101325.)\nH = thermo.get_enthalpy(temperature=298.15, verbose=False)\nS = thermo.get_entropy(temperature=298.15, pressure=101325., verbose=False)\n\nprint(f"H Gibbs energy: {G}")\nprint(f"H Enthalpy: {H}")\nprint(f"H Entropy: {S}")\n\n# Save results\nresults = {\'gibbs_energy\': G, \'enthalpy\': H, \'entropy\': S}\nwith open(\'h_thermo.json\', \'w\') as f:\n    json.dump(results, f)\nEOF\n\nMPLBACKEND=agg python calc_thermo_h.py\n# answer for H+ (Hint: consider the \'geometry\' parameter and set vib_energies to an empty list [])\n# Load h_results.json for the H+ energy\n# Save results to h_thermo.json\nno longer need to OPEN all that crap\n'' returned non-zero exit status 127.

At this point you should have three dictionaries _h\_results_, _oh\_results_ and _h2o\_results_ that includes the electronic energy, enthalpy and entropy of these species.

__Calculate the change in electronic energy $\Delta$E, the change in enthalpy $\Delta$H, the entropy contribution T$\Delta$S, and the Gibbs free energy change $\Delta$G for the reaction (Eqns 23-27):__

H$_2$O $⟶$ OH$^{-}$ + H$^+$


In [47]:
import json
with open ('h_results.json', 'r') as f:
    h_results = json.load(f)
with open ('geomopt_.json', 'r') as f:
    geomopt_results = json.load(f)

# answer for deltaE
# Load results from h_results.json, geomopt_results.json

FileNotFoundError: [Errno 2] No such file or directory: 'geomopt_.json'

In [44]:
deltaE= h_results['energy'] - geomopt_results['h2o_energy']
print (deltaE)

NameError: name 'h_results' is not defined

In [82]:
import json
with open('h_thermo.json', 'r') as f:
    h_thermo = json.load(f)
with open('oh_thermo.json', 'r') as f:
    oh_thermo = json.load(f)
with open('h2o_thermo.json', 'r') as f:
    h2o_thermo = json.load(f)

    print(h_thermo['enthalpy'])
    print(oh_thermo['enthalpy'])
    print(h2o_thermo['enthalpy'] )
# answer for deltaH
# Load results from h_thermo.json, oh_thermo.json, h2o_thermo.json

0.0642314260010328
-2062.2161180831718
-2078.8745120820067


H enthalpy = 0.0642314260010328 eV
OH enthalpy =-2062.2161180831718 eV
H2O enthalpy = -2078.8745120820067 eV



In [58]:
import json
with open('h_thermo.json', 'r') as f:
    h_thermo = json.load(f)
with open('oh_thermo.json', 'r') as f:
    oh_thermo = json.load(f)
with open('h2o_thermo.json', 'r') as f:
    h2o_thermo = json.load(f)

    print(h_thermo['entropy'])
    print(oh_thermo['entropy'])
    print(h2o_thermo['entropy'])


# answer for deltaS
# Load results from h_thermo.json, oh_thermo.json, h2o_thermo.json

0.0011281009772134308
0.0017270314504990998
0.0020548830397835263


H entropy = 0.0011281009772134308 eV
OH entropy =0.0017270314504990998 eV
H2O entropy = 0.0020548830397835263 eV



In [59]:
import json
with open('h_thermo.json', 'r') as f:
    h_thermo = json.load(f)
with open('oh_thermo.json', 'r') as f:
    oh_thermo = json.load(f)
with open('h2o_thermo.json', 'r') as f:
    h2o_thermo = json.load(f)

    print(h_thermo['gibbs_energy'])
    print(oh_thermo['gibbs_energy'])
    print(h2o_thermo['gibbs_energy'])
# answer for deltaG
# Load results from h_thermo.json, oh_thermo.json, h2o_thermo.json

-0.27211188035515155
-2062.731032510138
-2079.487175460318


H gibbs free energy = -0.27211188035515155 eV
OH gibbs free energy =-2062.731032510138 eV
H20 gibbs free energy =-2079.487175460318 eV

In [67]:
deltaE= (h_thermo['gibbs_energy'] + oh_thermo['gibbs_energy'] - h2o_thermo['gibbs_energy'])
print(deltaE)

16.48403106982505


rxn gibbs free energy = 16.48403106982505 eV

delta E in eV

In [66]:
deltaH= (h_thermo['enthalpy'] + oh_thermo['enthalpy'] - h2o_thermo['enthalpy'])
print(deltaH)

16.722625424836224


In [65]:
deltaS= (h_thermo['entropy'] + oh_thermo['entropy'] - h2o_thermo['entropy'])
print(deltaS)

0.0008002493879290043


__Calculate K$_a$ and pK$_a$ at 298.15 K for the acid-ionisation reaction H2O ⇌ H$^+$ + OH$^-$ (Eqns 1 and 2 of this preamble). Discuss how this compares with the experimentally determined value and why it might differ.__

In [76]:
import math

G = deltaE * (96485)
T = 298.15
value = -(G // (8.314472 * T))
Ka = math.exp(value)
print(Ka)


4.142257567365285e-279


In [72]:
pka = -math.log10(Ka)
print(pka)

278.3827628999844


calculated delta G: 1590.47 kJ/ mol, theoretical literature value: 1605.57 kj/mol

delta G was calculated with a 0.94% percentage error. This suggests that the computational approach was accurate.

Ka for this reaction is incredibly low, suggesting that this reaction does not occur readily, and that the equilibrium for this reaction is to the lesft hand side.

This computational method utilised b3lyp, which is a sufficient tool for dft. Alternatively, other programs could be used such as b2plyp.

b3lyp is known to be an efficient calculation method, and is known to be able to opperate significantly faster than other methods. This allows b3lyp to render large molecules with ease.

On the other hand, b2plyp is known to have a higher accuracy than for b3lyp, especially for the computation of non- covalent interactions. b2plyp is also more effective at calculating anharmonic vibrational frequencies and electronic excitation energies. b3lyp does not account for dispersion forces.

In this case, b3lyp was sufficient for undergoing DFT calulations for this relatively small system, however, b2plyp would likely be more accurate for this application.

lit source
https://webbook.nist.gov/cgi/cbook.cgi?ID=C7732185&Mask=1#Thermo-Gas
